src/gui/instrument/EnvelopeGraph.cpp

```diff
@@ -54,102 +56,208 @@ void EnvelopeGraph::modelChanged()
 
-	if (x5 > 174)
+		float preliminarySegmentWidth = 80. / 182. * availableWidth;
+		
+		const float neededWidth = sumOfSegments * preliminarySegmentWidth;
+		
+		if (neededWidth > availableWidth && sumOfSegments != 0.)
+		{
+			return computeMaximumSegmentWidthRelative();
+		}
+		else
+		{
+			return preliminarySegmentWidth;
+		}
+	};
+
+	// This is the maximum width that each of the five segments (DAHDR) can occupy.
+	float maximumSegmentWidth;
+
+	switch (m_scaling)
 	{
-		x1 = (x1 * 174) / x5;
-		x2 = (x2 * 174) / x5;
-		x3 = (x3 * 174) / x5;
-		x4 = (x4 * 174) / x5;
-		x5 = (x5 * 174) / x5;
+		case ScalingMode::Absolute:
+			maximumSegmentWidth = computeMaximumSegmentWidthAbsolute();
+			break;
+		case ScalingMode::Relative:
+		{
+			maximumSegmentWidth = computeMaximumSegmentWidthRelative();
+			break;
+		}
+		case ScalingMode::Dynamic:
+		default:
+		{
+			maximumSegmentWidth = computeMaximumSegmentWidthDynamic();
+			break;
+		}
 	}
-	x1 += 2;
-	x2 += 2;
-	x3 += 2;
-	x4 += 2;
-	x5 += 2;
-
-	p.drawLine(x1, y_base, x2, y_base - avail_height);
-	p.fillRect(x1 - 1, y_base - 2, 4, 4, end_points_bg_color);
-	p.fillRect(x1, y_base - 1, 2, 2, end_points_color);
-
-	p.drawLine(x2, y_base - avail_height, x3, y_base - avail_height);
-	p.fillRect(x2 - 1, y_base - 2 - avail_height, 4, 4,
-							end_points_bg_color);
-	p.fillRect(x2, y_base - 1 - avail_height, 2, 2, end_points_color);
-
-	const int sustainHeight = static_cast<int>(y_base - avail_height + (1 - sustain) * avail_height);
-
-	p.drawLine(x3, y_base-avail_height, x4, sustainHeight);
-	p.fillRect(x3 - 1, y_base - 2 - avail_height, 4, 4, end_points_bg_color);
-	p.fillRect(x3, y_base - 1 - avail_height, 2, 2, end_points_color);
-	
-	p.drawLine(x4, sustainHeight, x5, y_base);
-	p.fillRect(x4 - 1, sustainHeight - 2, 4, 4, end_points_bg_color);
-	p.fillRect(x4, sustainHeight - 1, 2, 2, end_points_color);
-	p.fillRect(x5 - 1, y_base - 2, 4, 4, end_points_bg_color);
-	p.fillRect(x5, y_base - 1, 2, 2, end_points_color);
+
+	// Compute the actual widths that the segments occupy and add them to the
+	// previous x coordinates starting at the margin.
+	const float predelayX = margin + predelayPercentage * maximumSegmentWidth;
+	const float attackX = predelayX + attackPercentage * maximumSegmentWidth;
+	const float holdX = attackX + holdPercentage * maximumSegmentWidth;
+	const float decayX = holdX + (decayPercentage * (1 - sustain)) * maximumSegmentWidth;
+	const float releaseX = decayX + releasePercentage * maximumSegmentWidth;
+
+	// Now compute the "full" points including y coordinates
+	const int yTop = 3;
+	const qreal yBase = height() - 3;
+	const int availableHeight = yBase - yTop;
+
+	const QPointF predelayPoint{predelayX, yBase};
+	const QPointF attackPoint{attackX, yTop};
+	const QPointF holdPoint{holdX, yTop};
+	const QPointF decayPoint{decayX, yTop + (1 - sustain) * availableHeight};
+	const QPointF releasePoint{releaseX, yBase};
+
+
```

Comment:
Single newline is fine (and others below)
```suggestion

```

Reply:
I have also kept these because they are intended to somewhat separate the three logical blocks:
1. Calculate values.
2. Draw lines.
3. Draw markers.

Reply:
Yes, I also thought about this but in this case I'd have to define structs/classes to forward the needed data into the functions so I rather kept it in one method even if it gets a bit longer that way.

Anyway, I have removed the second newlines with commit 26d260e7650.